transcom/mymove
# [MB-1299] Implementing payment request lookup for requested pickup date

## Description

This PR adds the functionality to look up the requested pickup date when a payment request is made.

## Reviewer Notes

Because the lookup needs to access the database, I changed some of the calling code to pass the database connection through.

This is our first implementation of a payment request lookup function, so please consider if there's anything here that you'd like changed before others use this as a possible model for future lookups.

## Setup

### Testing Dev

1. First run `make db_dev_e2e_populate` to ensure your database is setup with the E2E data.
1. Next, save this example JSON payload (which references E2E UUIDs) to a file called `pr_payload_dev.json`:
    ```json
    {
      "isFinal": false,
      "moveTaskOrderID": "5d4b25bb-eb04-4c03-9a81-ee0398cb779e",
      "serviceItems": [
        {
          "id": "9db1bf43-0964-44ff-8384-3297951f6781"
        }
      ]
    }
    ```
    Note that the service item ID above is referencing the service `DLH` (domestic linehaul), which has `RequestedPickupDate` as an expected parameter.
1. Ensure your server is running: `make server_run`
1. Run this command to fetch the "available to prime" MTOs and look at the payment request state for the `moveTaskOrderID` above in the returned JSON: `go run ./cmd/prime-api-client --insecure fetch-mtos | jq`
1. Look at the `payment_service_item_params` table and you should not see any records.
1. Now let's create a payment request using our payload above.  There are two different ways to do this:
    - With the `--filename` flag: `go run ./cmd/prime-api-client --insecure create-payment-request --filename pr_payload_dev.json | jq`
    - From stdin: `go run ./cmd/prime-api-client --insecure create-payment-request - < pr_payload_dev.json | jq`
1. Fetch the MTOs again and verify the payment request(s) you created are there: `go run ./cmd/prime-api-client --insecure fetch-mtos | jq`
1. Look at the `payment_service_item_params` table and verify that there are now params that include `RequestedPickupDate ` and that the value is the requested pickup date from the `mto_shipment`.
1. This query will also show you the key/value param pairs for an MTO service item id:
    ```sql
    select key, value from payment_service_item_params
    inner join payment_service_items psi on payment_service_item_params.payment_service_item_id = psi.id
    inner join service_item_param_keys sipk on payment_service_item_params.service_item_param_key_id = sipk.id
    where mto_service_item_id = '9db1bf43-0964-44ff-8384-3297951f6781';
    ```

## Code Review Verification Steps


## References

* [Jira story](https://dp3.atlassian.net/browse/MB-1299) for this change
